gateway/src/lib.rs

```diff
@@ -787,6 +797,101 @@ impl Actor {
+    ///
+    /// Each cron checkpoint will be checked against each other using blake hashing.
+    fn submit_cron(rt: &mut impl Runtime, params: CronCheckpoint) -> Result<RawBytes, ActorError> {
+        // submit cron can only be performed by signable addresses
+        rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
+
+        let msgs = rt.transaction(|st: &mut State, rt| {
+            // first we check the epoch is the correct one
+            let genesis_epoch = st.genesis_epoch;
+
+            // we process only it's multiple of cron_period since genesis_epoch
+            if (params.epoch - genesis_epoch) % st.cron_period != 0 {
+                return Err(actor_error!(illegal_argument, "epoch not allowed"));
+            }
+
+            let store = rt.store();
+            let submitter = rt.message().caller();
+
+            st.cron_submissions
+                .modify(store, |hamt| {
+                    let epoch_key = BytesKey::from(params.epoch.to_be_bytes().as_slice());
+                    let mut submission = match hamt.get(&epoch_key)? {
+                        Some(s) => s.clone(),
+                        None => CronSubmission::new(store)?,
+                    };
+
+                    if submission.is_executed() {
+                        return Err(anyhow!("epoch already executed"));
+                    }
+
+                    let epoch = params.epoch;
+                    let execution_status = submission.submit(store, submitter, params)?;
+
+                    if st.last_cron_executed_epoch + st.cron_period != epoch {
+                        // there are pending epoch to be executed,
+                        // just store the submission and skip execution
+                        hamt.set(epoch_key, submission)?;
+                        return Ok(None);
+                    }
+
+                    match execution_status {
+                        VoteExecutionStatus::ThresholdNotReached
+                        | VoteExecutionStatus::ReachingConsensus => {
+                            // threshold or consensus not reached, store submission and return
+                            hamt.set(epoch_key, submission)?;
+                            return Ok(None);
+                        }
+                        VoteExecutionStatus::RoundAbort => {
+                            submission.abort(store)?;
+                            hamt.set(epoch_key, submission)?;
+                            return Ok(None);
+                        }
+                        VoteExecutionStatus::ConsensusReached => {}
+                    }
+
+                    st.last_cron_executed_epoch = epoch;
+
+                    let msgs = submission
+                        .load_most_submitted_checkpoint(store)?
+                        .unwrap()
+                        .top_down_msgs;
+                    submission.executed();
+                    hamt.set(epoch_key, submission)?;
+
```

Comment:
We should remove from here the submission to clean it from the storage, right?

Reply:
oh, we don't keep historical submissions? Yeah, that works too, we have the whole state, can ask client to track.